File: 133-clone-graph/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    unordered_map<Node*,Node*> umap;
    Node* cloneGraph(Node* node) {
        if(node==NULL) return NULL;
        if(umap.find(node)!=umap.end()) return umap[node];
        Node* newnode=new Node(node->val);
        umap[node]=newnode;
        for(auto nei:node->neighbors){
            newnode->neighbors.push_back(cloneGraph(nei));
        }
        return newnode;
    }
};

File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    HashMap<Node,Node> umap=new HashMap<>();
    public Node cloneGraph(Node node) {
        if(node==null) return node;
        if(umap.containsKey(node)) return umap.get(node);
        Node newnode=new Node(node.val);
        umap.put(node,newnode);
        for(Node nei:node.neighbors){
            newnode.neighbors.add(cloneGraph(nei));
        }
        
        return newnode;
    }
}